In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
from sklearn.preprocessing import LabelEncoder
from sklearn.model_selection import train_test_split
from sklearn.metrics import accuracy_score, classification_report, confusion_matrix,roc_auc_score,roc_curve
from sklearn.linear_model import LogisticRegression
from sklearn.tree import DecisionTreeClassifier
from sklearn.ensemble import RandomForestClassifier, GradientBoostingClassifier, AdaBoostClassifier
from sklearn.svm import SVC, OneClassSVM
from sklearn.neighbors import KNeighborsClassifier
from sklearn.naive_bayes import GaussianNB
from xgboost import XGBClassifier
from catboost import CatBoostClassifier
from lightgbm import LGBMClassifier
from sklearn.ensemble import StackingClassifier
from sklearn.linear_model import RidgeClassifier, Lasso
from sklearn.preprocessing import StandardScaler,FunctionTransformer
from sklearn.compose import   ColumnTransformer
import sys
import os
from logger import logger
from imblearn.over_sampling import SMOTE
from utils import evaluate_model
from sklearn.metrics import accuracy_score, f1_score, classification_report, roc_auc_score
os.environ["JOBLIB_MULTIPROCESSING"] = "0"
os.environ["LOKY_MAX_CPU_COUNT"] = "1"
import warnings
warnings.filterwarnings('ignore')

## Load preprocessed datasets

In [2]:
try:
    train_df = pd.read_csv('../data/preprocessed/train_df.csv')
    test_df = pd.read_csv('../data/preprocessed/test_df.csv')
    print("Train shape:", train_df.shape)
    print("Test shape:", test_df.shape)
except Exception as e:
    print(f'Error occured {e}')


Train shape: (254750, 23)
Test shape: (63688, 23)


### Split Features and Target

In [3]:
x_train=train_df.drop(columns='Stay')
y_train=train_df['Stay']
x_test=test_df.drop(columns='Stay')
y_test=test_df['Stay']

print("x_train shape:", x_train.shape)
print("y_train shape:", y_train.shape)
print("x_test shape:", x_test.shape)
print("y_test shape:", y_test.shape)

x_train shape: (254750, 22)
y_train shape: (254750,)
x_test shape: (63688, 22)
y_test shape: (63688,)


In [ ]:
try:
    logger.info("Applying SMOTE to balance training data...")
    smote =SMOTE(sampling_strategy='auto',random_state=42,k_neighbors=10)
    x_resampled, y_resampled = smote.fit_resample(x_train, y_train)
    logger.info(f"SMOTE applied. Resampled x shape: {x_resampled.shape}, y shape: {y_resampled.shape}")
except Exception as e:
    logger.error(f" Error during SMOTE resampling: {e}")
    raise

In [ ]:
knn_pipeline = Pipeline([
    ('scaler', StandardScaler()),
    ('pca', PCA(n_components=10)),
    ('knn', KNeighborsClassifier(n_neighbors=5, n_jobs=-1))
])

base_models = {
    "Logistic Regression": LogisticRegression(max_iter=300, class_weight='balanced', n_jobs=-1),
    "Decision Tree": DecisionTreeClassifier(class_weight='balanced', max_depth=10),
    "Random Forest": RandomForestClassifier(class_weight='balanced', n_jobs=-1, n_estimators=100, max_depth=10),
    "KNN": knn_pipeline,
    "Naive Bayes": GaussianNB(),
    # "SVM": SVC(kernel='linear', probability=True, class_weight='balanced')
}

ensemble_models = {
    "XGBoost": XGBClassifier(use_label_encoder=False, eval_metric='mlogloss', tree_method='hist', max_depth=6, n_estimators=100, n_jobs=-1),
    "LightGBM": LGBMClassifier(class_weight='balanced', max_depth=6, n_estimators=100, n_jobs=-1),
    "CatBoost": CatBoostClassifier(verbose=0, auto_class_weights='Balanced', iterations=100, thread_count=1),
    "Stacking Ensemble": StackingClassifier(
        estimators=[
            ('lr', LogisticRegression(class_weight='balanced', n_jobs=-1)),
            ('xgb', XGBClassifier(use_label_encoder=False, eval_metric='mlogloss', n_estimators=50, n_jobs=-1)),
        ],
        final_estimator=LogisticRegression(class_weight='balanced', n_jobs=-1),
        cv=3,
        n_jobs=-1
    )
}
results = []

def timed_evaluate(name, model, x_train, y_train, x_test, y_test):
    start = time.time()
    logger.info(f" Training started: {name}")
    try:
        result = evaluate_model(name, model, x_train, y_train, x_test, y_test)
        end = time.time()
        logger.info(f" Training complete: {name} (Time: {end - start:.2f}s)")
        return result
    except Exception as e:
        logger.error(f" Training failed: {name} — {str(e)}")
        return None

def train_models(model_dict, label):
    logger.info(f" Starting evaluation of {label} models...")
    for name, model in model_dict.items():
        result = timed_evaluate(name, model, x_resampled, y_resampled, x_test, y_test)
        if result:
            results.append(result)

try:
    train_models(base_models, "Base")
    train_models(ensemble_models, "Ensemble")

    # Save model performance summary
    results_df = pd.DataFrame(results)
    results_df.sort_values(by="Test Accuracy", ascending=False, inplace=True)

    logger.info(" Model training completed. Summary of results:")
    logger.info(f"\n{results_df.to_string(index=False)}")

    print("\n Final Model Comparison:\n")
    print(results_df)

except Exception as e:
    logger.critical(" Critical failure during model evaluation", exc_info=True)



🔍 Training model: Logistic Regression
 Train Accuracy: 0.2639
 Test Accuracy: 0.2269
 F1-score: 0.2429
 AUC-ROC: 0.7247
 Classification Report:
               precision    recall  f1-score   support

           0       0.17      0.56      0.26      4689
           1       0.40      0.21      0.27     15561
           2       0.45      0.27      0.34     17603
           3       0.24      0.09      0.13     10981
           4       0.06      0.11      0.07      2357
           5       0.31      0.20      0.24      7128
           6       0.01      0.15      0.02       554
           7       0.09      0.06      0.07      2031
           8       0.10      0.30      0.15       941
           9       0.04      0.24      0.07       552
          10       0.27      0.48      0.35      1291

    accuracy                           0.23     63688
   macro avg       0.19      0.24      0.18     63688
weighted avg       0.32      0.23      0.24     63688


🔍 Training model: Decision Tree
 Train A

In [20]:
#Saving the models perforamnce results :
results_df.to_csv(os.path.join(os.path.join(os.path.dirname(os.getcwd()),'data','preprocessed'),'results_df.csv'),index=False)

In [ ]:
all_models = {**base_models, **ensemble_models}

# List of models with Test Accuracy > 0.35
top_models = results_df[results_df['Test Accuracy']>0.35]['Model'].to_list()

# Filter the models
best_models = {name: model for name, model in all_models.items() if name in top_models}

# Display final filtered models
print(" Filtered Best Models (Test Accuracy > 0.35):")
for name in best_models:
    print(f"• {name}")
best_models

✅ Filtered Best Models (Test Accuracy > 0.35):
• Decision Tree
• Random Forest
• XGBoost
• LightGBM
• CatBoost
• Stacking Ensemble


{'Decision Tree': DecisionTreeClassifier(class_weight='balanced', max_depth=10),
 'Random Forest': RandomForestClassifier(class_weight='balanced', max_depth=10, n_jobs=-1),
 'XGBoost': XGBClassifier(base_score=None, booster=None, callbacks=None,
               colsample_bylevel=None, colsample_bynode=None,
               colsample_bytree=None, device=None, early_stopping_rounds=None,
               enable_categorical=False, eval_metric='mlogloss',
               feature_types=None, feature_weights=None, gamma=None,
               grow_policy=None, importance_type=None,
               interaction_constraints=None, learning_rate=None, max_bin=None,
               max_cat_threshold=None, max_cat_to_onehot=None,
               max_delta_step=None, max_depth=6, max_leaves=None,
               min_child_weight=None, missing=nan, monotone_constraints=None,
               multi_strategy=None, n_estimators=100, n_jobs=-1,
               num_parallel_tree=None, ...),
 'LightGBM': LGBMClassifier(

In [29]:
import optuna
import joblib
from sklearn.metrics import accuracy_score, f1_score
import optuna.visualization as vis
from IPython.display import display
import logging
from logger import logger

models = {
    "Decision Tree": DecisionTreeClassifier(class_weight='balanced'),
    "Random Forest": RandomForestClassifier(class_weight='balanced'),
    "XGBoost": XGBClassifier(use_label_encoder=False, eval_metric='mlogloss', tree_method='hist'),
    "LightGBM": LGBMClassifier(class_weight='balanced'),
    "CatBoost": CatBoostClassifier(verbose=0, auto_class_weights='Balanced'),
}

# === Objective Function ===
def objective(trial, model_name, x_train, y_train, x_test, y_test):
    try:
        if model_name == "Decision Tree":
            params = {
                "max_depth": trial.suggest_categorical("max_depth", [3, 5, 10]),
                "min_samples_split": trial.suggest_categorical("min_samples_split", [5, 10]),
                "min_samples_leaf": trial.suggest_categorical("min_samples_leaf", [1, 2, 5]),
                "criterion": trial.suggest_categorical("criterion", ["gini", "entropy"]),
                "random_state": trial.suggest_int("random_state", 40, 44)
            }
        elif model_name == "Random Forest":
            params = {
                "n_estimators": trial.suggest_categorical("n_estimators", [50, 100]),
                "max_depth": trial.suggest_categorical("max_depth", [5,10, 20]),
                "min_samples_split": trial.suggest_categorical("min_samples_split", [2, 5, 10]),
                "min_samples_leaf": trial.suggest_categorical("min_samples_leaf", [2, 4]),
                "max_features": trial.suggest_categorical("max_features", [ "sqrt", "log2"]),
                # "bootstrap": trial.suggest_categorical("bootstrap", [True, False]),
                "random_state": trial.suggest_categorical("random_state", list(42,44)),
                "class_weight": "balanced"
            }
        elif model_name == "XGBoost":
            params = {
                "n_estimators": trial.suggest_categorical("n_estimators", [50, 100]),
                "learning_rate": trial.suggest_categorical("learning_rate", [0.01, 0.1]),
                "max_depth": trial.suggest_categorical("max_depth", [5, 10]),
                "min_child_weight": trial.suggest_categorical("min_child_weight", [3, 5]),
                "subsample": trial.suggest_categorical("subsample", [0.6, 0.8]),
                "colsample_bytree": trial.suggest_categorical("colsample_bytree", [0.6, 0.8]),
                "gamma": trial.suggest_categorical("gamma", [0, 0.1, 0.2]),
                "reg_alpha": trial.suggest_categorical("reg_alpha", [ 0.01, 0.1]),
                "reg_lambda": trial.suggest_categorical("reg_lambda", [ 1.5, 2]),
                "random_state": trial.suggest_categorical("random_state", [42, 44]),
                "use_label_encoder": False,
                "eval_metric": "mlogloss",
                "tree_method": "hist",
                "n_jobs": 1
            }
        elif model_name == "LightGBM":
            params = {
                "n_estimators": trial.suggest_int("n_estimators", 50, 200, step=50),
                "learning_rate": trial.suggest_float("learning_rate", 0.01, 0.3),
                "max_depth": trial.suggest_int("max_depth", 3, 15),
                "num_leaves": trial.suggest_int("num_leaves", 20, 100, step=10),
                "min_child_samples": trial.suggest_int("min_child_samples", 5, 20),
                "subsample": trial.suggest_float("subsample", 0.6, 1.0),
                "colsample_bytree": trial.suggest_float("colsample_bytree", 0.6, 1.0),
                "class_weight": "balanced",
                "n_jobs": 1
            }
        elif model_name == "CatBoost":
            params = {
                "iterations": trial.suggest_int("iterations", 50, 200, step=50),
                "learning_rate": trial.suggest_float("learning_rate", 0.01, 0.3),
                "depth": trial.suggest_int("depth", 4, 10),
                "l2_leaf_reg": trial.suggest_float("l2_leaf_reg", 1.0, 10.0),
                "auto_class_weights": "Balanced",
                "verbose": 0,
                "thread_count": 1
            }
        else:
            raise ValueError("Model not supported.")

        model = models[model_name].__class__(**params)
        model.fit(x_train, y_train)
        y_pred = model.predict(x_test)
        return f1_score(y_test, y_pred, average='weighted')
    
    except Exception as e:
        logger.error(f"Failed during trial for {model_name}: {str(e)}")
        return 0.0


In [31]:
def objective(trial, model_name, x_train, y_train, x_test, y_test):
    try:
        if model_name == "Decision Tree":
            params = {
                "max_depth": trial.suggest_categorical("max_depth", [3, 5, None]),
                "min_samples_split": trial.suggest_categorical("min_samples_split", [2, 5]),
                "min_samples_leaf": trial.suggest_categorical("min_samples_leaf", [1, 2]),
                "criterion": trial.suggest_categorical("criterion", ["gini", "entropy"]),
                "random_state": 42
            }

        elif model_name == "Random Forest":
            params = {
                "n_estimators": trial.suggest_categorical("n_estimators", [50, 100]),
                "max_depth": trial.suggest_categorical("max_depth", [None, 10]),
                "min_samples_split": 2,
                "min_samples_leaf": 1,
                "max_features": "sqrt",
                "bootstrap": True,
                "random_state": 42,
                "class_weight": "balanced"
            }

        elif model_name == "XGBoost":
            params = {
                "n_estimators": trial.suggest_categorical("n_estimators", [50, 100]),
                "learning_rate": trial.suggest_categorical("learning_rate", [0.05, 0.1]),
                "max_depth": trial.suggest_categorical("max_depth", [3, 5]),
                "subsample": 0.8,
                "colsample_bytree": 0.8,
                "random_state": 42,
                "use_label_encoder": False,
                "eval_metric": "mlogloss",
                "tree_method": "hist",
                "n_jobs": 1
            }

        elif model_name == "LightGBM":
            params = {
                "n_estimators": trial.suggest_categorical("n_estimators", [50, 100]),
                "learning_rate": trial.suggest_float("learning_rate", 0.05, 0.2),
                "max_depth": trial.suggest_int("max_depth", 3, 7),
                "num_leaves": trial.suggest_int("num_leaves", 20, 50, step=10),
                "subsample": 0.8,
                "colsample_bytree": 0.8,
                "class_weight": "balanced",
                "n_jobs": 1
            }

        elif model_name == "CatBoost":
            params = {
                "iterations": trial.suggest_categorical("iterations", [50, 100]),
                "learning_rate": trial.suggest_float("learning_rate", 0.05, 0.2),
                "depth": trial.suggest_int("depth", 4, 6),
                "auto_class_weights": "Balanced",
                "verbose": 0,
                "thread_count": 1
            }

        else:
            raise ValueError("Model not supported.")

        model = models[model_name].__class__(**params)
        model.fit(x_train, y_train)
        y_pred = model.predict(x_test)
        return f1_score(y_test, y_pred, average='weighted')
    
    except Exception as e:
        logger.error(f"Failed during trial for {model_name}: {str(e)}")
        return 0.0


In [ ]:
results = []
tuned_models = {}
best_params_=[]

for model_name in models:
    try:
        logger.info(f"Tuning model: {model_name}")
        study = optuna.create_study(direction="maximize")
        study.optimize(lambda trial: objective(trial, model_name, x_resampled, y_resampled, x_test, y_test), n_trials=15)

        logger.info(f"Best params for {model_name}: {study.best_params}")
        best_params = study.best_params
        best_params_.append({'Model': model_name,'best_params': best_params})

        final_model = models[model_name].__class__(**best_params)
        final_model.fit(x_resampled, y_resampled)
        y_pred = final_model.predict(x_test)

        acc = accuracy_score(y_test, y_pred)
        f1 = f1_score(y_test, y_pred, average='weighted')

        logger.info(f"{model_name} - Accuracy: {acc:.4f}, F1-score: {f1:.4f}")

        # joblib.dump(final_model, f"{model_name.replace(' ', '_').lower()}_best_model.pkl")
        # tuned_models[model_name] = final_model

        # Visualize
        # display(vis.plot_optimization_history(study))
        # display(vis.plot_param_importances(study))

        results.append({"Model": model_name, "Accuracy": acc, "F1-score": f1})

    except Exception as e:
        logger.error(f"Error tuning {model_name}: {str(e)}")
#Summary of all model performances :
summary_df = pd.DataFrame(results).sort_values(by="F1-score", ascending=False)
logger.info("Final Model Performance Summary:")
display(summary_df)


[I 2025-04-21 09:59:45,197] A new study created in memory with name: no-name-f959fb02-b8df-4599-b9e3-3cead1a2b97f
[I 2025-04-21 09:59:51,392] Trial 0 finished with value: 0.03443388460838167 and parameters: {'max_depth': 3, 'min_samples_split': 5, 'min_samples_leaf': 2, 'criterion': 'gini'}. Best is trial 0 with value: 0.03443388460838167.
[I 2025-04-21 10:00:43,298] Trial 1 finished with value: 0.30616649290143017 and parameters: {'max_depth': None, 'min_samples_split': 5, 'min_samples_leaf': 2, 'criterion': 'entropy'}. Best is trial 1 with value: 0.30616649290143017.
[I 2025-04-21 10:00:52,644] Trial 2 finished with value: 0.26724060042319714 and parameters: {'max_depth': 5, 'min_samples_split': 5, 'min_samples_leaf': 1, 'criterion': 'gini'}. Best is trial 1 with value: 0.30616649290143017.
[I 2025-04-21 10:01:02,119] Trial 3 finished with value: 0.26724060042319714 and parameters: {'max_depth': 5, 'min_samples_split': 2, 'min_samples_leaf': 2, 'criterion': 'gini'}. Best is trial 1 w

[LightGBM] [Warning] Found whitespace in feature_names, replace with underlines
[LightGBM] [Info] Auto-choosing row-wise multi-threading, the overhead of testing was 0.057283 seconds.
You can set `force_row_wise=true` to remove the overhead.
And if memory is not enough, you can set `force_col_wise=true`.
[LightGBM] [Info] Total Bins 5609
[LightGBM] [Info] Number of data points in the train set: 768768, number of used features: 22
[LightGBM] [Info] Start training from score -2.397895
[LightGBM] [Info] Start training from score -2.397895
[LightGBM] [Info] Start training from score -2.397895
[LightGBM] [Info] Start training from score -2.397895
[LightGBM] [Info] Start training from score -2.397895
[LightGBM] [Info] Start training from score -2.397895
[LightGBM] [Info] Start training from score -2.397895
[LightGBM] [Info] Start training from score -2.397895
[LightGBM] [Info] Start training from score -2.397895
[LightGBM] [Info] Start training from score -2.397895
[LightGBM] [Info] Start tr

[I 2025-04-21 11:43:51,317] Trial 0 finished with value: 0.35533377302381675 and parameters: {'n_estimators': 50, 'learning_rate': 0.14081625918882507, 'max_depth': 3, 'num_leaves': 20}. Best is trial 0 with value: 0.35533377302381675.


[LightGBM] [Warning] Found whitespace in feature_names, replace with underlines
[LightGBM] [Info] Auto-choosing row-wise multi-threading, the overhead of testing was 0.045781 seconds.
You can set `force_row_wise=true` to remove the overhead.
And if memory is not enough, you can set `force_col_wise=true`.
[LightGBM] [Info] Total Bins 5609
[LightGBM] [Info] Number of data points in the train set: 768768, number of used features: 22
[LightGBM] [Info] Start training from score -2.397895
[LightGBM] [Info] Start training from score -2.397895
[LightGBM] [Info] Start training from score -2.397895
[LightGBM] [Info] Start training from score -2.397895
[LightGBM] [Info] Start training from score -2.397895
[LightGBM] [Info] Start training from score -2.397895
[LightGBM] [Info] Start training from score -2.397895
[LightGBM] [Info] Start training from score -2.397895
[LightGBM] [Info] Start training from score -2.397895
[LightGBM] [Info] Start training from score -2.397895
[LightGBM] [Info] Start tr

[I 2025-04-21 11:44:23,391] Trial 1 finished with value: 0.38581741427579835 and parameters: {'n_estimators': 50, 'learning_rate': 0.1938355152910487, 'max_depth': 6, 'num_leaves': 50}. Best is trial 1 with value: 0.38581741427579835.


[LightGBM] [Warning] Found whitespace in feature_names, replace with underlines
[LightGBM] [Info] Auto-choosing row-wise multi-threading, the overhead of testing was 0.050598 seconds.
You can set `force_row_wise=true` to remove the overhead.
And if memory is not enough, you can set `force_col_wise=true`.
[LightGBM] [Info] Total Bins 5609
[LightGBM] [Info] Number of data points in the train set: 768768, number of used features: 22
[LightGBM] [Info] Start training from score -2.397895
[LightGBM] [Info] Start training from score -2.397895
[LightGBM] [Info] Start training from score -2.397895
[LightGBM] [Info] Start training from score -2.397895
[LightGBM] [Info] Start training from score -2.397895
[LightGBM] [Info] Start training from score -2.397895
[LightGBM] [Info] Start training from score -2.397895
[LightGBM] [Info] Start training from score -2.397895
[LightGBM] [Info] Start training from score -2.397895
[LightGBM] [Info] Start training from score -2.397895
[LightGBM] [Info] Start tr

[I 2025-04-21 11:45:33,673] Trial 2 finished with value: 0.38590461026496786 and parameters: {'n_estimators': 100, 'learning_rate': 0.06774923857990149, 'max_depth': 7, 'num_leaves': 50}. Best is trial 2 with value: 0.38590461026496786.


[LightGBM] [Warning] Found whitespace in feature_names, replace with underlines
[LightGBM] [Info] Auto-choosing row-wise multi-threading, the overhead of testing was 0.045927 seconds.
You can set `force_row_wise=true` to remove the overhead.
And if memory is not enough, you can set `force_col_wise=true`.
[LightGBM] [Info] Total Bins 5609
[LightGBM] [Info] Number of data points in the train set: 768768, number of used features: 22
[LightGBM] [Info] Start training from score -2.397895
[LightGBM] [Info] Start training from score -2.397895
[LightGBM] [Info] Start training from score -2.397895
[LightGBM] [Info] Start training from score -2.397895
[LightGBM] [Info] Start training from score -2.397895
[LightGBM] [Info] Start training from score -2.397895
[LightGBM] [Info] Start training from score -2.397895
[LightGBM] [Info] Start training from score -2.397895
[LightGBM] [Info] Start training from score -2.397895
[LightGBM] [Info] Start training from score -2.397895
[LightGBM] [Info] Start tr

[I 2025-04-21 11:45:59,662] Trial 3 finished with value: 0.3644851657505449 and parameters: {'n_estimators': 50, 'learning_rate': 0.10880205901655496, 'max_depth': 4, 'num_leaves': 40}. Best is trial 2 with value: 0.38590461026496786.


[LightGBM] [Warning] Found whitespace in feature_names, replace with underlines
[LightGBM] [Info] Auto-choosing row-wise multi-threading, the overhead of testing was 0.051684 seconds.
You can set `force_row_wise=true` to remove the overhead.
And if memory is not enough, you can set `force_col_wise=true`.
[LightGBM] [Info] Total Bins 5609
[LightGBM] [Info] Number of data points in the train set: 768768, number of used features: 22
[LightGBM] [Info] Start training from score -2.397895
[LightGBM] [Info] Start training from score -2.397895
[LightGBM] [Info] Start training from score -2.397895
[LightGBM] [Info] Start training from score -2.397895
[LightGBM] [Info] Start training from score -2.397895
[LightGBM] [Info] Start training from score -2.397895
[LightGBM] [Info] Start training from score -2.397895
[LightGBM] [Info] Start training from score -2.397895
[LightGBM] [Info] Start training from score -2.397895
[LightGBM] [Info] Start training from score -2.397895
[LightGBM] [Info] Start tr

[I 2025-04-21 11:46:56,635] Trial 4 finished with value: 0.38898370469761107 and parameters: {'n_estimators': 100, 'learning_rate': 0.12385470240706338, 'max_depth': 6, 'num_leaves': 30}. Best is trial 4 with value: 0.38898370469761107.


[LightGBM] [Warning] Found whitespace in feature_names, replace with underlines
[LightGBM] [Info] Auto-choosing row-wise multi-threading, the overhead of testing was 0.043025 seconds.
You can set `force_row_wise=true` to remove the overhead.
And if memory is not enough, you can set `force_col_wise=true`.
[LightGBM] [Info] Total Bins 5609
[LightGBM] [Info] Number of data points in the train set: 768768, number of used features: 22
[LightGBM] [Info] Start training from score -2.397895
[LightGBM] [Info] Start training from score -2.397895
[LightGBM] [Info] Start training from score -2.397895
[LightGBM] [Info] Start training from score -2.397895
[LightGBM] [Info] Start training from score -2.397895
[LightGBM] [Info] Start training from score -2.397895
[LightGBM] [Info] Start training from score -2.397895
[LightGBM] [Info] Start training from score -2.397895
[LightGBM] [Info] Start training from score -2.397895
[LightGBM] [Info] Start training from score -2.397895
[LightGBM] [Info] Start tr

[I 2025-04-21 11:47:17,561] Trial 5 finished with value: 0.34180300089941035 and parameters: {'n_estimators': 50, 'learning_rate': 0.08867665367723769, 'max_depth': 3, 'num_leaves': 50}. Best is trial 4 with value: 0.38898370469761107.


[LightGBM] [Warning] Found whitespace in feature_names, replace with underlines
[LightGBM] [Info] Auto-choosing row-wise multi-threading, the overhead of testing was 0.045945 seconds.
You can set `force_row_wise=true` to remove the overhead.
And if memory is not enough, you can set `force_col_wise=true`.
[LightGBM] [Info] Total Bins 5609
[LightGBM] [Info] Number of data points in the train set: 768768, number of used features: 22
[LightGBM] [Info] Start training from score -2.397895
[LightGBM] [Info] Start training from score -2.397895
[LightGBM] [Info] Start training from score -2.397895
[LightGBM] [Info] Start training from score -2.397895
[LightGBM] [Info] Start training from score -2.397895
[LightGBM] [Info] Start training from score -2.397895
[LightGBM] [Info] Start training from score -2.397895
[LightGBM] [Info] Start training from score -2.397895
[LightGBM] [Info] Start training from score -2.397895
[LightGBM] [Info] Start training from score -2.397895
[LightGBM] [Info] Start tr

[I 2025-04-21 11:48:18,667] Trial 6 finished with value: 0.38837540460239933 and parameters: {'n_estimators': 100, 'learning_rate': 0.1070974565607821, 'max_depth': 6, 'num_leaves': 50}. Best is trial 4 with value: 0.38898370469761107.


[LightGBM] [Warning] Found whitespace in feature_names, replace with underlines
[LightGBM] [Info] Auto-choosing row-wise multi-threading, the overhead of testing was 0.046928 seconds.
You can set `force_row_wise=true` to remove the overhead.
And if memory is not enough, you can set `force_col_wise=true`.
[LightGBM] [Info] Total Bins 5609
[LightGBM] [Info] Number of data points in the train set: 768768, number of used features: 22
[LightGBM] [Info] Start training from score -2.397895
[LightGBM] [Info] Start training from score -2.397895
[LightGBM] [Info] Start training from score -2.397895
[LightGBM] [Info] Start training from score -2.397895
[LightGBM] [Info] Start training from score -2.397895
[LightGBM] [Info] Start training from score -2.397895
[LightGBM] [Info] Start training from score -2.397895
[LightGBM] [Info] Start training from score -2.397895
[LightGBM] [Info] Start training from score -2.397895
[LightGBM] [Info] Start training from score -2.397895
[LightGBM] [Info] Start tr

[I 2025-04-21 11:48:43,194] Trial 7 finished with value: 0.364274241550689 and parameters: {'n_estimators': 50, 'learning_rate': 0.10346109893386353, 'max_depth': 4, 'num_leaves': 30}. Best is trial 4 with value: 0.38898370469761107.


[LightGBM] [Warning] Found whitespace in feature_names, replace with underlines
[LightGBM] [Info] Auto-choosing row-wise multi-threading, the overhead of testing was 0.044920 seconds.
You can set `force_row_wise=true` to remove the overhead.
And if memory is not enough, you can set `force_col_wise=true`.
[LightGBM] [Info] Total Bins 5609
[LightGBM] [Info] Number of data points in the train set: 768768, number of used features: 22
[LightGBM] [Info] Start training from score -2.397895
[LightGBM] [Info] Start training from score -2.397895
[LightGBM] [Info] Start training from score -2.397895
[LightGBM] [Info] Start training from score -2.397895
[LightGBM] [Info] Start training from score -2.397895
[LightGBM] [Info] Start training from score -2.397895
[LightGBM] [Info] Start training from score -2.397895
[LightGBM] [Info] Start training from score -2.397895
[LightGBM] [Info] Start training from score -2.397895
[LightGBM] [Info] Start training from score -2.397895
[LightGBM] [Info] Start tr

[I 2025-04-21 11:49:35,818] Trial 8 finished with value: 0.38774134519746617 and parameters: {'n_estimators': 100, 'learning_rate': 0.19979142208406492, 'max_depth': 7, 'num_leaves': 30}. Best is trial 4 with value: 0.38898370469761107.


[LightGBM] [Warning] Found whitespace in feature_names, replace with underlines
[LightGBM] [Info] Auto-choosing row-wise multi-threading, the overhead of testing was 0.044951 seconds.
You can set `force_row_wise=true` to remove the overhead.
And if memory is not enough, you can set `force_col_wise=true`.
[LightGBM] [Info] Total Bins 5609
[LightGBM] [Info] Number of data points in the train set: 768768, number of used features: 22
[LightGBM] [Info] Start training from score -2.397895
[LightGBM] [Info] Start training from score -2.397895
[LightGBM] [Info] Start training from score -2.397895
[LightGBM] [Info] Start training from score -2.397895
[LightGBM] [Info] Start training from score -2.397895
[LightGBM] [Info] Start training from score -2.397895
[LightGBM] [Info] Start training from score -2.397895
[LightGBM] [Info] Start training from score -2.397895
[LightGBM] [Info] Start training from score -2.397895
[LightGBM] [Info] Start training from score -2.397895
[LightGBM] [Info] Start tr

[I 2025-04-21 11:50:26,432] Trial 9 finished with value: 0.38800851008882276 and parameters: {'n_estimators': 100, 'learning_rate': 0.18508277164718479, 'max_depth': 6, 'num_leaves': 20}. Best is trial 4 with value: 0.38898370469761107.


[LightGBM] [Warning] Found whitespace in feature_names, replace with underlines
[LightGBM] [Info] Auto-choosing row-wise multi-threading, the overhead of testing was 0.045691 seconds.
You can set `force_row_wise=true` to remove the overhead.
And if memory is not enough, you can set `force_col_wise=true`.
[LightGBM] [Info] Total Bins 5609
[LightGBM] [Info] Number of data points in the train set: 768768, number of used features: 22
[LightGBM] [Info] Start training from score -2.397895
[LightGBM] [Info] Start training from score -2.397895
[LightGBM] [Info] Start training from score -2.397895
[LightGBM] [Info] Start training from score -2.397895
[LightGBM] [Info] Start training from score -2.397895
[LightGBM] [Info] Start training from score -2.397895
[LightGBM] [Info] Start training from score -2.397895
[LightGBM] [Info] Start training from score -2.397895
[LightGBM] [Info] Start training from score -2.397895
[LightGBM] [Info] Start training from score -2.397895
[LightGBM] [Info] Start tr

[I 2025-04-21 11:51:19,977] Trial 10 finished with value: 0.3879201835581478 and parameters: {'n_estimators': 100, 'learning_rate': 0.15152190979683122, 'max_depth': 5, 'num_leaves': 40}. Best is trial 4 with value: 0.38898370469761107.


[LightGBM] [Warning] Found whitespace in feature_names, replace with underlines
[LightGBM] [Info] Auto-choosing row-wise multi-threading, the overhead of testing was 0.045308 seconds.
You can set `force_row_wise=true` to remove the overhead.
And if memory is not enough, you can set `force_col_wise=true`.
[LightGBM] [Info] Total Bins 5609
[LightGBM] [Info] Number of data points in the train set: 768768, number of used features: 22
[LightGBM] [Info] Start training from score -2.397895
[LightGBM] [Info] Start training from score -2.397895
[LightGBM] [Info] Start training from score -2.397895
[LightGBM] [Info] Start training from score -2.397895
[LightGBM] [Info] Start training from score -2.397895
[LightGBM] [Info] Start training from score -2.397895
[LightGBM] [Info] Start training from score -2.397895
[LightGBM] [Info] Start training from score -2.397895
[LightGBM] [Info] Start training from score -2.397895
[LightGBM] [Info] Start training from score -2.397895
[LightGBM] [Info] Start tr

[I 2025-04-21 11:52:16,579] Trial 11 finished with value: 0.389372488117666 and parameters: {'n_estimators': 100, 'learning_rate': 0.12916963413279267, 'max_depth': 6, 'num_leaves': 30}. Best is trial 11 with value: 0.389372488117666.


[LightGBM] [Warning] Found whitespace in feature_names, replace with underlines
[LightGBM] [Info] Auto-choosing row-wise multi-threading, the overhead of testing was 0.046943 seconds.
You can set `force_row_wise=true` to remove the overhead.
And if memory is not enough, you can set `force_col_wise=true`.
[LightGBM] [Info] Total Bins 5609
[LightGBM] [Info] Number of data points in the train set: 768768, number of used features: 22
[LightGBM] [Info] Start training from score -2.397895
[LightGBM] [Info] Start training from score -2.397895
[LightGBM] [Info] Start training from score -2.397895
[LightGBM] [Info] Start training from score -2.397895
[LightGBM] [Info] Start training from score -2.397895
[LightGBM] [Info] Start training from score -2.397895
[LightGBM] [Info] Start training from score -2.397895
[LightGBM] [Info] Start training from score -2.397895
[LightGBM] [Info] Start training from score -2.397895
[LightGBM] [Info] Start training from score -2.397895
[LightGBM] [Info] Start tr

[I 2025-04-21 11:53:26,313] Trial 12 finished with value: 0.38796568300974515 and parameters: {'n_estimators': 100, 'learning_rate': 0.14804887887914484, 'max_depth': 5, 'num_leaves': 30}. Best is trial 11 with value: 0.389372488117666.


[LightGBM] [Warning] Found whitespace in feature_names, replace with underlines
[LightGBM] [Info] Auto-choosing row-wise multi-threading, the overhead of testing was 0.067628 seconds.
You can set `force_row_wise=true` to remove the overhead.
And if memory is not enough, you can set `force_col_wise=true`.
[LightGBM] [Info] Total Bins 5609
[LightGBM] [Info] Number of data points in the train set: 768768, number of used features: 22
[LightGBM] [Info] Start training from score -2.397895
[LightGBM] [Info] Start training from score -2.397895
[LightGBM] [Info] Start training from score -2.397895
[LightGBM] [Info] Start training from score -2.397895
[LightGBM] [Info] Start training from score -2.397895
[LightGBM] [Info] Start training from score -2.397895
[LightGBM] [Info] Start training from score -2.397895
[LightGBM] [Info] Start training from score -2.397895
[LightGBM] [Info] Start training from score -2.397895
[LightGBM] [Info] Start training from score -2.397895
[LightGBM] [Info] Start tr

[I 2025-04-21 11:54:52,535] Trial 13 finished with value: 0.38436388055018084 and parameters: {'n_estimators': 100, 'learning_rate': 0.1305941574553452, 'max_depth': 5, 'num_leaves': 30}. Best is trial 11 with value: 0.389372488117666.


[LightGBM] [Warning] Found whitespace in feature_names, replace with underlines
[LightGBM] [Info] Auto-choosing row-wise multi-threading, the overhead of testing was 0.067973 seconds.
You can set `force_row_wise=true` to remove the overhead.
And if memory is not enough, you can set `force_col_wise=true`.
[LightGBM] [Info] Total Bins 5609
[LightGBM] [Info] Number of data points in the train set: 768768, number of used features: 22
[LightGBM] [Info] Start training from score -2.397895
[LightGBM] [Info] Start training from score -2.397895
[LightGBM] [Info] Start training from score -2.397895
[LightGBM] [Info] Start training from score -2.397895
[LightGBM] [Info] Start training from score -2.397895
[LightGBM] [Info] Start training from score -2.397895
[LightGBM] [Info] Start training from score -2.397895
[LightGBM] [Info] Start training from score -2.397895
[LightGBM] [Info] Start training from score -2.397895
[LightGBM] [Info] Start training from score -2.397895
[LightGBM] [Info] Start tr

[I 2025-04-21 11:56:00,952] Trial 14 finished with value: 0.38732419391075174 and parameters: {'n_estimators': 100, 'learning_rate': 0.1668363798598973, 'max_depth': 6, 'num_leaves': 20}. Best is trial 11 with value: 0.389372488117666.


[LightGBM] [Warning] Found whitespace in feature_names, replace with underlines
[LightGBM] [Info] Auto-choosing row-wise multi-threading, the overhead of testing was 0.045470 seconds.
You can set `force_row_wise=true` to remove the overhead.
And if memory is not enough, you can set `force_col_wise=true`.
[LightGBM] [Info] Total Bins 5609
[LightGBM] [Info] Number of data points in the train set: 768768, number of used features: 22
[LightGBM] [Info] Start training from score -2.397895
[LightGBM] [Info] Start training from score -2.397895
[LightGBM] [Info] Start training from score -2.397895
[LightGBM] [Info] Start training from score -2.397895
[LightGBM] [Info] Start training from score -2.397895
[LightGBM] [Info] Start training from score -2.397895
[LightGBM] [Info] Start training from score -2.397895
[LightGBM] [Info] Start training from score -2.397895
[LightGBM] [Info] Start training from score -2.397895
[LightGBM] [Info] Start training from score -2.397895
[LightGBM] [Info] Start tr

[I 2025-04-21 11:56:57,325] A new study created in memory with name: no-name-471ee2ba-ef50-4de3-bfeb-57bc89f78d5d
[I 2025-04-21 11:57:55,583] Trial 0 finished with value: 0.3409065559563314 and parameters: {'iterations': 50, 'learning_rate': 0.09361089613218576, 'depth': 5}. Best is trial 0 with value: 0.3409065559563314.
[I 2025-04-21 11:59:32,057] Trial 1 finished with value: 0.3604220304417195 and parameters: {'iterations': 100, 'learning_rate': 0.12664780040321869, 'depth': 4}. Best is trial 1 with value: 0.3604220304417195.
[I 2025-04-21 12:00:23,267] Trial 2 finished with value: 0.31853100332588896 and parameters: {'iterations': 50, 'learning_rate': 0.07748763934095868, 'depth': 4}. Best is trial 1 with value: 0.3604220304417195.
[I 2025-04-21 12:01:18,678] Trial 3 finished with value: 0.3616583172528475 and parameters: {'iterations': 50, 'learning_rate': 0.18535511207866817, 'depth': 5}. Best is trial 3 with value: 0.3616583172528475.
[I 2025-04-21 12:02:15,020] Trial 4 finished

0:	learn: 2.2236035	total: 506ms	remaining: 50.1s
1:	learn: 2.1275940	total: 1.03s	remaining: 50.7s
2:	learn: 2.0581611	total: 1.53s	remaining: 49.6s
3:	learn: 2.0063576	total: 2.03s	remaining: 48.8s
4:	learn: 1.9612135	total: 2.53s	remaining: 48.1s
5:	learn: 1.9239463	total: 3.03s	remaining: 47.5s
6:	learn: 1.8943147	total: 3.54s	remaining: 47.1s
7:	learn: 1.8711303	total: 4.05s	remaining: 46.5s
8:	learn: 1.8517239	total: 4.56s	remaining: 46.1s
9:	learn: 1.8345588	total: 5.07s	remaining: 45.7s
10:	learn: 1.8206395	total: 5.58s	remaining: 45.2s
11:	learn: 1.8011511	total: 6.11s	remaining: 44.8s
12:	learn: 1.7881338	total: 6.65s	remaining: 44.5s
13:	learn: 1.7761817	total: 7.17s	remaining: 44s
14:	learn: 1.7659726	total: 7.68s	remaining: 43.5s
15:	learn: 1.7560506	total: 8.19s	remaining: 43s
16:	learn: 1.7476227	total: 8.73s	remaining: 42.6s
17:	learn: 1.7371512	total: 9.25s	remaining: 42.2s
18:	learn: 1.7280839	total: 9.78s	remaining: 41.7s
19:	learn: 1.7200038	total: 10.3s	remaining: 

Model  Accuracy  F1-score
3       LightGBM  0.416798  0.387380
4       CatBoost  0.407769  0.379285
2        XGBoost  0.405178  0.376608
1  Random Forest  0.388236  0.373408
0  Decision Tree  0.306306  0.306545

In [36]:
best_models_=[
    ('Decision Tree', DecisionTreeClassifier(max_depth= 10,min_samples_split= 10,min_samples_leaf=2,criterion='gini',random_state= 46)),
    ('Random Forest', RandomForestClassifier(class_weight='balanced', max_depth=10,n_estimators= 100,min_samples_split=2, n_jobs=-1,max_features='sqrt',bootstrap=True,random_state=42)),
    ( 'XGBoost',XGBClassifier(n_estimators=100,learning_rate=0.1,max_depth=5,subsample=0.8,colsample_bytree=0.8,random_state=42,use_label_encoder=False,eval_metric='mlogloss',tree_method='hist',n_jobs=-1)),

]
final_estimator=XGBClassifier(n_estimators=100,learning_rate=0.1,max_depth=5,subsample=0.8,colsample_bytree=0.8,random_state=42,use_label_encoder=False,eval_metric='mlogloss',tree_method='hist',n_jobs=-1)

In [38]:
try:
    logger.info("Training Stacking Ensemble with tuned models...")

    stacking_model = StackingClassifier(
        estimators=best_models_,
        final_estimator=final_estimator,
        n_jobs=-1
    )
    logger.info(" Stacking model fitting with tuned models...")
    stacking_model.fit(x_resampled, y_resampled)
    logger.info(" Stacking model training completed and going to predict ...")
    y_pred_stack = stacking_model.predict(x_test)

    acc_stack = accuracy_score(y_test, y_pred_stack)
    f1_stack = f1_score(y_test, y_pred_stack, average='weighted')
    logging.info(f"Stacking Ensemble - Accuracy: {acc_stack:.4f}, F1-score: {f1_stack:.4f}")

    # joblib.dump(stacking_model, "stacking_ensemble_best_model.pkl")
    results.append({"Model": "Stacking Ensemble", "Accuracy": acc_stack, "F1-score": f1_stack})

except Exception as e:
    logging.error(f"Error training Stacking Ensemble: {str(e)}")

In [46]:
joblib.dump(LGBMClassifier(n_estimators=100,learning_rate=0.12916,max_depth=6,num_leaves=30,subsample=0.8,colsample_bytree=0.8,class_weight='balanced', n_jobs=-1),"../models/best_model.pkl")

['../models/best_model.pkl']

In [50]:
summary_df.to_dict()

{'Model': {3: 'LightGBM',
  4: 'CatBoost',
  2: 'XGBoost',
  1: 'Random Forest',
  0: 'Decision Tree',
  5: 'Stacking Ensemble'},
 'Accuracy': {3: 0.41679751287526695,
  4: 0.407769124481849,
  2: 0.4051783695515639,
  1: 0.38823640246200225,
  0: 0.3063057404848637,
  5: 0.4106582087677427},
 'F1-score': {3: 0.38738004912868207,
  4: 0.379285172425095,
  2: 0.3766075312215814,
  1: 0.3734079146880604,
  0: 0.3065450353347443,
  5: 0.37888590116663196}}

In [ ]:
#Saving the ensemble model performance in summary_df :
summary_df.loc[len(summary_df),:]=results[-1]

In [43]:
summary_df

Model  Accuracy  F1-score
3           LightGBM  0.416798  0.387380
4           CatBoost  0.407769  0.379285
2            XGBoost  0.405178  0.376608
1      Random Forest  0.388236  0.373408
0      Decision Tree  0.306306  0.306545
5  Stacking Ensemble  0.410658  0.378886

In [49]:
model=joblib.load('../models/best_model.pkl')
model

LGBMClassifier(class_weight='balanced', colsample_bytree=0.8,
               learning_rate=0.12916, max_depth=6, n_jobs=-1, num_leaves=30,
               subsample=0.8)

##  Model Comparison Report :

### 1. Objective

The goal of this task was to evaluate various machine learning models for predicting patient length of hospital stay and recommend the best-performing model for production deployment.

### 2. Evaluation Metrics

All models were evaluated using the following metrics:

- **Training Accuracy**
- **Test Accuracy**
- **F1-Score**
- **AUC-ROC Score**

### 3. Performance Before Hyperparameter Tuning

| Model                | Train Accuracy | Test Accuracy | F1-Score | AUC-ROC |
|----------------------|----------------|----------------|----------|---------|
| XGBoost              | 0.5504         | 0.4179         | 0.3896   | 0.7977  |
| LightGBM             | 0.4952         | 0.4162         | 0.3863   | 0.7959  |
| CatBoost             | 0.4963         | 0.4151         | 0.3877   | 0.7957  |
| Stacking Ensemble    | 0.4908         | 0.3983         | 0.3757   | 0.7881  |
| Random Forest        | 0.4282         | 0.3729         | 0.3512   | 0.7692  |
| Decision Tree        | 0.3914         | 0.3658         | 0.3496   | 0.7514  |
| K-Nearest Neighbors  | 0.7476         | 0.2408         | 0.2545   | 0.6004  |
| Logistic Regression  | 0.2639         | 0.2269         | 0.2429   | 0.7247  |
| Naive Bayes          | 0.2058         | 0.1347         | 0.1407   | 0.6763  |

**Observation:**
- XGBoost, LightGBM, and CatBoost consistently outperformed others across all metrics.
- KNN exhibited severe overfitting.
- Simpler models like Logistic Regression and Naive Bayes underperformed.

### 4. Performance After Hyperparameter Tuning

| Model              | Test Accuracy | F1-Score |
|--------------------|---------------|----------|
| LightGBM           | 0.4168        | 0.3874   |
| CatBoost           | 0.4078        | 0.3793   |
| XGBoost            | 0.4052        | 0.3766   |
| Stacking Ensemble  | 0.4107        | 0.3789   |
| Random Forest      | 0.3882        | 0.3734   |
| Decision Tree      | 0.3063        | 0.3065   |

**Observation:**
- Hyperparameter tuning led to slight improvements in all models.
- LightGBM emerged as the best performer, followed by CatBoost and XGBoost.
- Decision Tree continued to show poor generalization performance even after tuning.

### 5. Recommended Model

Based on the comparison of models before and after hyperparameter tuning, the **LightGBM Classifier** is recommended for production deployment due to its:

- High test accuracy and F1-score
- Balanced performance and generalization
- Compatibility with large tabular datasets
- Fast training speed and interpretability



##  Report on Data Challenges and Solutions :

### 1. Missing Values

**Problem:**  
The dataset contained missing values in several columns.

**Solution:**  
- Used **SimpleImputer** with strategy `most_frequent` for categorical variables and `mean` for numerical variables.
- Justification: Most frequent imputation helps maintain categorical distributions, while mean imputation works well for numeric skewed data.

### 2. Imbalanced Target Variable

**Problem:**  
The target variable (`Length of Stay`) was highly imbalanced, with shorter stays being far more frequent.

**Solution:**  
- Applied **class weights = 'balanced'** in classifiers like LightGBM and RandomForest.
- Also experimented with **SMOTE** for oversampling minority classes during model tuning.


### 3. High Cardinality in Categorical Features

**Problem:**  
Columns like `Department`,`Type_of_Admission` and `Hospital Code` had a large number of unique categories.

**Solution:**  
- Used **Target Encoding** for high cardinality features.
- Used **One-Hot Encoding** for low cardinality features to retain interpretability.



### 4.  Feature Scaling

**Problem:**  
Some models like KNN and Logistic Regression were sensitive to feature scales.

**Solution:**  
- Applied **StandardScaler** to normalize numerical features.
- Tree-based models were excluded from scaling as they are scale-invariant.

### 6.  Model Interpretability

**Problem:**  
Need to explain predictions to stakeholders in the healthcare domain.

**Solution:**  
- Selected **LightGBM** for its interpretability.
- Used **SHAP values** and **feature importance plots** to highlight contributing features.


### 7. Long Training Times

**Problem:**  
Training time was high for ensemble models.

**Solution:**  
- Reduced feature space using feature importance.
- Enabled **multi-threading** with `n_jobs = -1` wherever supported.


#### This structured approach helped in building a reliable and interpretable model for predicting patient hospital stay durations. All challenges were addressed with appropriate preprocessing, feature engineering, and modeling techniques.
